### PR TITLE
Adding scenario tests for WCF #1402

### PR DESCRIPTION
**Pull request by KKhurin**
Note: a couple of negative tests will fail until the feature is complete.

### CODE REVIEW

**Comment by KKhurin** [comment]
@dotnet-bot test outerloop Windows_NT please

**Comment by KKhurin** [comment]
@dotnet-bot test OuterLoop Selfhost Windows please

**Comment by KKhurin** [comment]
@dotnet-bot test OuterLoop Windows_NT please

File: src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/SessionTests.cs [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+
+            // *** CLEANUP *** \\
+            factory.Close();
+            ((ICommunicationObject)channel).Close();
+        }
+        finally
+        {
+            ScenarioTestHelpers.CloseCommunicationObjects((ICommunicationObject)channel, factory);
+        }
+//throw new Exception("");
```
Comment (zhenlan) [inline src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/SessionTests.cs]: NIT: clean up :)

File: src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/SessionTests.cs [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+                });
+            }
+            finally
+            {
+                ScenarioTestHelpers.CloseCommunicationObjects((ICommunicationObject)channel1);
+            }
+        }
+    }
+
+    [ActiveIssue(1402)]
```
Comment (KKhurin) [inline src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/SessionTests.cs]: change attribute

**Comment by KKhurin** [comment]
@dotnet-bot test outerloop Windows_NT please

**Comment by StephenBonikowsky** [comment]
LGTM
We've been having issues with OSX since we updated to version 10.12, should have nothing to do with your changes, don't block on OSX10.12 CI failures.

File: src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/SessionTests.cs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+using System.Threading.Tasks;
+using Infrastructure.Common;
+using Xunit;
+
+public static class SessionTests
+{
+    // A basic test to verify that a session gets created and terminated
+    [WcfFact]
+    [OuterLoop]
+    static void Test_IsInitiating_IsTerminating()
```
Comment (shmao) [inline src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/SessionTests.cs]: Nit: specify the visibility explicitly.

File: src/System.Private.ServiceModel/tools/IISHostedWcfService/App_code/SessionTestsService.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+using System;
+using System.ServiceModel;
+using System.Threading;
+
+namespace WcfService
+{
+    [ServiceBehavior(InstanceContextMode = InstanceContextMode.PerSession)]
+    public class SessionTestsDefaultService : ISessionTestsDefaultService
+    {
+        private int _AValue = 0, _BValue = 0;
```
Comment (shmao) [inline src/System.Private.ServiceModel/tools/IISHostedWcfService/App_code/SessionTestsService.cs]: Nit: use `_camelCase` for private fields

File: src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/SessionTests.cs [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+            Assert.Equal(dataC1.MethodAValue, A1);
+            Assert.Equal(dataC1.MethodBValue, B1);
+            Assert.Equal(dataC2.MethodAValue, A2);
+            Assert.Equal(dataC2.MethodBValue, B2);
+
+            // Methods A & B update the private A & B fields in the service  
+            // and return the original values of the service field instance
+            // Our implementation of IService1 has InstanceContextMode = InstanceContextMode.PerSession
+            // so all dataA&B* should have the original values == 0
+            Assert.Equal(dataA1 & dataB1 & dataA2 & dataB2, 0);
```
Comment (shmao) [inline src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/SessionTests.cs]: Should we assert the value of  `dataA1 | dataB1 | dataA2 | dataB2` instead? Besides, as `0` is the expected value it should be the first parameter.

File: src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/SessionTests.cs [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+                
+                (channel as ICommunicationObject).Open();
+                // The following Non Initiating method can be called without calling an Initiating method first 
+                // if the session was implicitly created by explicitly calling channel.Open()
+                channel.MethodBNonInitiating(0xB);
+
+                // The service behind service2Url has the same contract and implementation as service1Url
+                // But it has a different binding with a very short receiveTimeout ="00:00:02"
+                // So waiting for just 5 seconds is enough to get the connection and the session implicitly closed
+                Task.Delay(8080).Wait();
```
Comment (shmao) [inline src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/SessionTests.cs]: Question: why `8080`? You said '5 seconds is enough'.

File: src/System.Private.ServiceModel/tools/IISHostedWcfService/App_code/testhosts/TcpSessionTestServiceHost.cs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+namespace WcfService
+{
+    public class TcpSessionTestServiceHostFactory : ServiceHostFactory
+    {
+        protected override ServiceHost CreateServiceHost(Type serviceType, Uri[] baseAddresses)
+        {
+            TcpSessionTestServiceHost serviceHost = new TcpSessionTestServiceHost(serviceType, baseAddresses);
+            return serviceHost;
+        }
+    }
```
Comment (shmao) [inline src/System.Private.ServiceModel/tools/IISHostedWcfService/App_code/testhosts/TcpSessionTestServiceHost.cs]: Nit: an empty line after `}`.

File: src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/SessionTests.cs (hunk 5 above)
Comment (KKhurin) [inline src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/SessionTests.cs]: Hmm... good catch! And perhaps for consistency I should revert the order. Thanks!

File: src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/SessionTests.cs (hunk 6 above)
Comment (KKhurin) [inline src/System.Private.ServiceModel/tests/Scenarios/Client/ClientBase/SessionTests.cs]: Artifacts from debugging something. Thanks for catching this.

**Pull request merged**